dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTinyVectorIndexWriter.cpp

```diff
@@ -0,0 +1,234 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <IO/Compression/CompressedWriteBuffer.h>
+#include <Storages/DeltaMerge/ColumnFile/ColumnFileTiny.h>
+#include <Storages/DeltaMerge/ColumnFile/ColumnFileTinyReader.h>
+#include <Storages/DeltaMerge/ColumnFile/ColumnFileTinyVectorIndexWriter.h>
+#include <Storages/DeltaMerge/Index/VectorIndex.h>
+#include <Storages/DeltaMerge/dtpb/dmfile.pb.h>
+
+
+namespace DB::ErrorCodes
+{
+extern const int ABORTED;
+} // namespace DB::ErrorCodes
+
+namespace DB::DM
+{
+
+ColumnFileTinyVectorIndexWriter::LocalIndexBuildInfo ColumnFileTinyVectorIndexWriter::getLocalIndexBuildInfo(
+    const LocalIndexInfosSnapshot & index_infos,
+    const ColumnFilePersistedSetPtr & file_set)
+{
+    assert(index_infos != nullptr);
+
+    LocalIndexBuildInfo build;
+    build.indexes_to_build = std::make_shared<LocalIndexInfos>();
+    build.file_ids.reserve(file_set->getColumnFileCount());
+    for (const auto & file : file_set->getFiles())
+    {
+        auto * tiny_file = file->tryToTinyFile();
+        if (!tiny_file)
+            continue;
+
+        bool any_new_index_build = false;
+        for (const auto & index : *index_infos)
+        {
+            auto schema = tiny_file->getSchema();
+            assert(schema != nullptr);
+            // The ColumnFileTiny may be built before col_id is added. Skip build indexes for it
+            if (!schema->getColIdToOffset().contains(index.column_id))
+                continue;
+
+            // Index already built, skip
+            if (tiny_file->hasIndex(index.index_id))
+                continue;
+
+            any_new_index_build = true;
```

Comment:
Why it happens before duplicated index check(L64-L68)?

Reply:
For example, there are file_1, file_2, file_3, and index_1, index_2, index_3, and file_1 with index_1, file_2 with index_2, file_3 with index_3.

`any_new_index_build` means whether the file needs to build the index, so `any_new_index_build` should be `true` for all files in this case.

`indexes_to_build` contains all index information that will be built in the task, so `indexes_to_build` should contain information of all indexes in this case.